### PR TITLE
Initial Sf3 support

### PR DESCRIPTION
**Pull request by chives**

### CODE REVIEW

**Comment by norzechowicz** [comment]
this PR should be created against master branch, also I would create 3.0 release from master rather than next 2.x

**Comment by chives** [comment]
This PR is meant to be fully BC so there is no need to jump to 3.0 yet

**Comment by chives** [comment]
@fsi-open

**Comment by norzechowicz** [comment]
So you are 100% sure this PR and all dependencies you upgraded are totally BC? 
Releasing new version cost nothing, it does not affect anything and projects that are already using 2.0 will not suffer because of changes you made, like for example using `@dev` in dependencies is not something I would like to see in stable release. Leave 2.0 as is, close this PR, create it against master and anybody who would like to upgrade his project to SF 3+ will just change admin-bundle dependency version from `2.0.x` to `3.0.x`.

**Comment by l3l0** [comment]
It is better IMHO to add support for sf3 on separate branch, without this BC glue code. Is someone want to use version for sf3 then just use other version in composer. Such BC code can be pain in maintenance later.

**Comment by chives** [comment]
@l3l0 the plan was to introduce sf3 compatibility in 2.x branch and remove the BC code in 3.0 which would require sf3. This approach would allow to switch to sf3 by changing the dependency from symfony only in the project using admin bundle. If sf3 compatibility is going to be added in admin's 3.0 release then we'll probably drop sf2 compatibility at once in 3.0. I still consider 2.1 (or some later 2.x) version with sf2/sf3 compatibility as not as bad idea as you and @norzechowicz think.

**Comment by norzechowicz** [comment]
yea but whats the point of having something between? With your approach you gonna need to support following admin bundle instances:
- v 2.0.0 with symfony > 2.8
- v 2.1.0 with symfony > 3.x (with this extra migration code)
- v 3.0.0 with symfony > 3.x
 
and this is pointless and because of this migration code it might easily break.

**Comment by norzechowicz** [comment]
@chives what is missing here? Except using only stable dependencies (I'm looking now at datasource) but I assume it's only a temporary change.

**Comment by chives** [comment]
@norzechowicz I think this PR is perfectly OK if we want to maintain support for symfony 2.x and 3.x but if so, this could be merged into some admin-bundle's 2.x release (maybe 2.2)

**Comment by norzechowicz** [comment]
but what's the point of maintaining version 2.x with support for sf3? One extra version that you gonna need to support later (just look at version 1.0 which still has unresolved bc break, I don't think that adding new version will bring any benefits). I would merge that to the master, and keep support for sf3 as a feature of admin-bundle 3.0. As a developer working on a project that is using admin bundle I would rather like to upgrade from `2.0 -> 3.0` than `2.0 -> 2.? -> 2.? -> 3.0 `

**Comment by chives** [comment]
I really don't mind adding support for sf2 in admin-bundle 3.0. We also plan some features to add into 2.x branches before 3.0 and they probably will be also merged into master.

**Comment by norzechowicz** [comment]
Ehm now I'm lost

> I really don't mind adding support for sf2 in admin-bundle 3.0

but the support for sf2 is something that should stay even in master, right? Or maybe you would like to keep something like:

* admin-bundle 2.0 -> sf 2.x
* admin-bundle 3.0 -> sf 3.x

**Comment by chives** [comment]
I mean adding support for sf3 in admin 3 and still maintaining support for sf2 in admin 3

File: Form/FeatureHelper.php [hunk 1]
```diff
@@ -0,0 +1,39 @@
+<?php
+
+namespace FSi\Bundle\AdminBundle\Form;
+
+use Symfony\Component\Form\FormTypeInterface;
+
+/**
+ * @internal
+ */
+class FeatureHelper
```
Comment (norzechowicz) [inline Form/FeatureHelper.php]: `TypeResolver` sounds better

**Comment by chives** [comment]
@fsi-open

File: Form/TypeSolver.php [hunk 2]
```diff
@@ -0,0 +1,39 @@
...
+    /**
+     * Return FQCN form type or old style form type
+     *
+     * @param string $fqcnType
+     * @param string|FormTypeInterface $shortType
+     * @return string|FormTypeInterface
+     */
+    public static function getFormType($fqcnType, $shortType)
+    {
+        return class_exists('Symfony\Component\Form\Extension\Core\Type\RangeType') ? $fqcnType : $shortType;
```
Comment (szymach) [inline Form/TypeSolver.php]: Pretty arbitrary IMO, is there no other way to differentiate?

File: composer.json [hunk 3]
```diff
@@ -17,44 +17,46 @@
...
-        "symfony/twig-bundle": "^2.3",
+        "symfony/twig-bundle": "^2.3|^3.0",
         "fsi/datagrid-bundle" : "^1.2",
-        "fsi/datasource-bundle": "^1.2.6",
-        "fsi/datasource": "^1.2.8",
-        "fsi/data-indexer" : "0.9.*",
+        "fsi/datasource-bundle": "^1.3@dev",
+        "fsi/datasource": "^1.3@dev",
+        "fsi/data-indexer" : "^0.9.0",
         "fsi/property-observer" : "^0.9.1",
```
Comment (szymach) [inline composer.json]: Is this used by the bundle or one of it's dependencies?

Comment (chives) [inline composer.json]: Not sure. Anyway it's not the subject of this PR.

File: Form/TypeSolver.php (hunk 2 above)
Comment (chives) [inline Form/TypeSolver.php]: `RangeType` was introduced in symfony 2.8. AFAIK it's the only new feature which can indicate if `symfony/form` 2.8 or newer is installed

**Comment by chives** [comment]
@fsi-open

**Comment by szymach** [comment]
@rn0 @JarekW

**Comment by chives** [comment]
@rn0

**Pull request merged**